garrysmod/lua/includes/extensions/string.lua

```diff
@@ -1,92 +1,76 @@
+local error = error
+local isnumber = isnumber
+local tonumber = tonumber
+local tostring = tostring
 
-local string = string
-local math = math
+local table_concat = table.concat
 
---[[---------------------------------------------------------
-	Name: string.ToTable( string )
------------------------------------------------------------]]
-function string.ToTable( str )
+local math_floor = math.floor
+local math_Round = math.Round
+
+local string_sub = string.sub
+local string_find = string.find
+local string_gsub = string.gsub
+local string_match = string.match
+local string_format = string.format
+
+getmetatable( "" ).__index = function( str, key )
+	if ( isnumber( key ) ) then
+		return string_sub( str, key, key )
+	end
+
+	local val = string[ key ]
+
+	if ( val == nil ) then
+		error( "attempt to index a string value with bad key ('" .. tostring( key ) .. "' is not part of the string library)", 2 )
+	end
+
+	return val
+end
+
+local function string_ToTable( str )
 	local tbl = {}
 
-	for i = 1, string.len( str ) do
-		tbl[i] = string.sub( str, i, i )
+	for i = 1, #str do
+		tbl[i] = string_sub( str, i, i )
 	end
 
 	return tbl
 end
 
---[[---------------------------------------------------------
-	Name: string.JavascriptSafe( string )
-	Desc: Takes a string and escapes it for insertion in to a JavaScript string
------------------------------------------------------------]]
+string.ToTable = string_ToTable
+
 local javascript_escape_replacements = {
 	["\\"] = "\\\\",
-	["\0"] = "\\x00" ,
-	["\b"] = "\\b" ,
-	["\t"] = "\\t" ,
-	["\n"] = "\\n" ,
-	["\v"] = "\\v" ,
-	["\f"] = "\\f" ,
-	["\r"] = "\\r" ,
+	["\0"] = "\\x00",
+	["\b"] = "\\b",
+	["\t"] = "\\t",
+	["\n"] = "\\n",
+	["\v"] = "\\v",
+	["\f"] = "\\f",
+	["\r"] = "\\r",
 	["\""] = "\\\"",
 	["\'"] = "\\\'"
 }
 
 function string.JavascriptSafe( str )
-
-	str = str:gsub( ".", javascript_escape_replacements )
-
+	str = string_gsub( str, ".", javascript_escape_replacements )
 	-- U+2028 and U+2029 are treated as line separators in JavaScript, handle separately as they aren't single-byte
-	str = str:gsub( "\226\128\168", "\\\226\128\168" )
-	str = str:gsub( "\226\128\169", "\\\226\128\169" )
-
-	return str
-
-end
-
---[[---------------------------------------------------------
-	Name: string.PatternSafe( string )
-	Desc: Takes a string and escapes it for insertion in to a Lua pattern
------------------------------------------------------------]]
-local pattern_escape_replacements = {
-	["("] = "%(",
-	[")"] = "%)",
-	["."] = "%.",
-	["%"] = "%%",
-	["+"] = "%+",
-	["-"] = "%-",
-	["*"] = "%*",
-	["?"] = "%?",
-	["["] = "%[",
-	["]"] = "%]",
-	["^"] = "%^",
-	["$"] = "%$",
-	["\0"] = "%z"
-}
+	str = string_gsub( str, "\226\128\168", "\\\226\128\168" )
 
-function string.PatternSafe( str )
-	return ( str:gsub( ".", pattern_escape_replacements ) )
+	return string_gsub( str, "\226\128\169", "\\\226\128\169" )

```

Comment:
What if we combine these 2 gsub with capture?

Reply:
you mean like `string_gsub( str, "\226\128\16([89])", "\\\226\128\16%1" )`

Reply:
Yeah